lib/internal/modules/esm/get_format.js

```diff
@@ -74,7 +73,7 @@ function extname(url) {
  */
 function getFileProtocolModuleFormat(url, context, ignoreErrors) {
   const ext = extname(url);
-  if (ext === '.js') {
+  if (ext === '.js' || ext === '') {
```

Comment:
I think we’re aiming to change how extensionless files are handled _as entry points,_ not in general from any `import`.

Reply:
How come? Both ESM with other protocols and CJS have no problems with extensionless imports.

Reply:
It would be very weird to have a file parsed as ESM only some of the time, I much prefer the current approach.

Reply:
In the original implementation before it was limited, we supported extensionless per `"type"` only for entry points, not for any `import`, per https://github.com/nodejs/node/pull/31415#issue-552101138. We’ve never supported extensionless for `import`. I could search around for why there was the original limitation to make this only for entry points, but I’d bet there was a good reason. There’s also not really a use case for `import` of extensionless files.

Reply:
> We’ve never supported extensionless for `import`

In the comment you just linked, you say:

> support for extensionless files in import was added only last month, in https://github.com/nodejs/node/pull/31021.

Reply:
Okay so we didn’t _never_ support it, but it wasn’t part of the original implementation. It was briefly added and then reverted. Probably because of this: https://github.com/nodejs/node/pull/31021#pullrequestreview-334268101

> One of the major features of the ES module resolver is being able to support new file extensions in future, which is reliant on the fact that we always throw for unknown file extensions.
> 
> Giving special treatment to `isMain` was how we did this while ensuring compatibility with existing `bin`s.
> 
> Happy to flesh this out further, but I don’t think we should lose that extension property for the ESM resolver.

Reply:
I dug through a lot of old issues and consulted @guybedford to try to rediscover why we limited extensionless support to entry points in the initial implementation, and why we removed even that in https://github.com/nodejs/node/pull/31415. Basically, we wanted to preserve the ability for future extensionless Wasm entry points. And I think we still do; that goal hasn’t changed, and https://github.com/nodejs/node/pull/49540 is a viable way to achieve that goal. I think years ago we weren’t aware of or really thinking about the “disambiguate based on magic bytes” solution, and so we just locked down the design space in the hope that we’d figure something out eventually. And I think the magic bytes approach is that something, and so it can open up the ability for both extensionless ESM entry points and extensionless Wasm entry points.

As for `import` of extensionless files, there would be the similar question of how to disambiguate an `import` of an extensionless JavaScript file from an extensionless Wasm file. The simplest solution is probably to just take same approach that we use for entry points, looking at magic bytes. Currently `import` of Wasm is gated behind the separate flag `--experimental-wasm-modules`, so it doesn’t matter all that much.

But I think we _can_ support `import` of extensionless JavaScript. JavaScript can be the default for an `import` statement, where “extensionless = JavaScript with the format determined by `type`” for `import` like for entry point, with other potential formats needing a way to disambiguate. Those new types would either need explicit extensions (the most likely and obvious disambiguation) or they would need mandatory import attributes (like JSON) or they would need magic bytes; _some_ way to disambiguate them from JavaScript. But I think we can unlock `import` of extensionless JavaScript without much risk of that prohibiting future enhancements.